### PR TITLE
CT: Add BlockHash op

### LINKED ISSUE #459: CT: add specification of the BLOCKHASH instruction
The specification of the BLOCKHASH instruction in Tosca CT is missing.

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds the rules for BLOCKHASH operation, a new condition to enforce whether the stack top is within or out of the range of the current block number and tests for it.

This PR fixes #459

### CODE REVIEW

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
I added a list of changes I think that should be applied to this PR. Please see the comments.

However, I think the chosen approach of introducing two new conditions (`WithinBlockRange` and `OutOfBlockRange`) does not really produce the tests we would like to cover for this instruction -- at least their current implementation are not testing what is desired.

Ideally, we find a condition format that would produce test cases covering the set
```
Param(0) \in { 0, CB-257, CB-256, CB-255, CB-1, CB, CB+1, 2^32 + CB - 1, 2^256-1 }
```
where `CB` is the current block number.

Maybe you can think about something to improve this? I would have to look into this myself in more detail, but I do not have the time right now to do so.

File: go/ct/rlz/conditions.go [hunk 1]
```diff
@@ -719,3 +719,98 @@ func (c *hasNotSelfDestructed) GetTestValues() []TestValue {
...
+
+func (c *withinBlockRange) Restrict(generator *gen.StateGenerator) {
+	paramVariable := c.blockNumber.GetVariable()
+	c.blockNumber.BindTo(generator)
+	generator.AddBlockNumberInRange(paramVariable)
+}
+
+func (c *withinBlockRange) GetTestValues() []TestValue {
+	property := Property(c.String())
+	domain := uint64Domain{}
```
[the remark this answers is not in the document]
Comment (facuMH) [reply go/ct/rlz/conditions.go]: A new domain is added that samples on the edge cases for both in and out of range `[CurrentBlock-257, CurrentBlock+1]`, and this domain is used to generate the test values, hence a bool is not needed.

File: go/ct/rlz/conditions.go [hunk 2]
```diff
@@ -719,3 +719,98 @@ func (c *hasNotSelfDestructed) GetTestValues() []TestValue {
...
+	testValues := []TestValue{}
+	var blockNumberDomain BlockNumberRangeDomain
+	for _, value := range blockNumberDomain.Samples(0) {
+		testValues = append(testValues, NewTestValue(property, domain, value, restrict))
+	}
+	return testValues
+}
+
+func (c *inRange256FromCurrentBlock) String() string {
+	return fmt.Sprintf("blockRangeOffset(%v)", c.blockNumber)
```
Comment (HerbertJordan) [inline go/ct/rlz/conditions.go:774]: nit: "inBlockRangeOffset(%v)" or "blockNumber - 256 ≤ %v ∧ %v < blockNumber" or "%v ∈ [blockNumber - 256, blockNumber-1]" or "%v ∈ [blockNumber - 256, blockNumber)"

File: go/ct/rlz/conditions.go [hunk 3]
```diff
@@ -719,3 +719,98 @@ func (c *hasNotSelfDestructed) GetTestValues() []TestValue {
...
+	c.blockNumber.BindTo(generator)
+	generator.AddBlockNumberOffset(paramVariable, -1)
+}
+
+func (c *outOfRange256FromCurrentBlock) GetTestValues() []TestValue {
+	return InRange256FromCurrentBlock(c.blockNumber).GetTestValues()
+}
+
+func (c *outOfRange256FromCurrentBlock) String() string {
+	return InRange256FromCurrentBlock(c.blockNumber).String()
```
Comment (HerbertJordan) [inline go/ct/rlz/conditions.go:813]: This is wrong. It should be the opposite. Something like
```
"%v ∉ [blockNumber - 256, blockNumber)"
```

File: go/ct/rlz/conditions.go [hunk 4]
```diff
@@ -719,3 +719,98 @@ func (c *hasNotSelfDestructed) GetTestValues() []TestValue {
...
+		bottom = s.BlockContext.BlockNumber - 256
+	}
+	uintParam := paramBlockNumber.Uint64()
+	return uintParam >= s.BlockContext.BlockNumber || bottom > uintParam, nil
+}
+
+func (c *outOfRange256FromCurrentBlock) Restrict(generator *gen.StateGenerator) {
+	paramVariable := c.blockNumber.GetVariable()
+	c.blockNumber.BindTo(generator)
+	generator.AddBlockNumberOffset(paramVariable, -1)
```
Comment (HerbertJordan) [inline go/ct/rlz/conditions.go:805]: As in the case of the `InRang256FromCurrentBlock`, this is overly restrictive.

File: go/ct/rlz/conditions.go [hunk 5]
```diff
@@ -719,3 +719,98 @@ func (c *hasNotSelfDestructed) GetTestValues() []TestValue {
...
+	paramBlockNumber, err := c.blockNumber.Eval(s)
+	if err != nil {
+		return false, err
+	}
+	bottom := uint64(0)
+	if s.BlockContext.BlockNumber > 256 {
+		bottom = s.BlockContext.BlockNumber - 256
+	}
+	uintParam := paramBlockNumber.Uint64()
+	return uintParam >= s.BlockContext.BlockNumber || bottom > uintParam, nil
```
Comment (HerbertJordan) [inline go/ct/rlz/conditions.go:799]: This should be simplified by reusing `inRange256FromCurrentBlock`'s implementation of the `Check` function:
```
res, err := InRange256FromCurrentBlock(c.blockNumber).Check(s)
return !res, err
```

File: go/ct/rlz/conditions_test.go [hunk 6]
```diff
@@ -288,3 +288,55 @@ func TestHasSelfDestructedCondition_HasSelfDestructRestrictsGeneratedStateToBeSe
 		t.Error("generated state does not satisfy condition")
 	}
 }
+
+func TestValidRange256_Check(t *testing.T) {
```
Comment (HerbertJordan) [inline go/ct/rlz/conditions_test.go:292]: This is a very high-level end-to-end test for the new conditions. It would be good to test Check, Restrict, and GetTestValues independently as well.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
I think there are a range of improvements that should still be applied before this can be merged. Please have a look at individual comments.

File: go/ct/gen/block_context.go [hunk 7]
```diff
@@ -13,43 +13,85 @@
...
 
 func NewBlockContextGenerator() *BlockContextGenerator {
-	return &BlockContextGenerator{}
+	maxBlockNumber, _ := common.GetForkBlock(common.R99_UnknownNextRevision)
+	return &BlockContextGenerator{blockNumbers: NewRangeSolver(0, maxBlockNumber),
+		variablesOffsetConstraints: []variableOffset{}}
 }
 
-func (*BlockContextGenerator) Generate(rnd *rand.Rand, revision common.Revision) (st.BlockContext, error) {
+func isWithin256Range(offset int64, min, max uint64) bool {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply go/ct/gen/block_context.go:41]: mmmm I see the confusion. My interpretation was that `offset` is in the range `[min, max)`. For this case, I think `isInRange(offset, min, max)` or `isInRange(min, max, offset)` are good options.

File: go/ct/gen/block_context.go [hunk 8]
```diff
@@ -13,43 +13,85 @@
...
+			// both not in range of block number
+			if !(isWithin256Range(offsetConstraint.offset, min, blockNumber) &&
+				isWithin256Range(int64(assignment[offsetConstraint.variable].Uint64()), min, blockNumber)) ||
+				// or not both out of range from block number
+				!(!isWithin256Range(offsetConstraint.offset, min, blockNumber) &&
+					!isWithin256Range(int64(assignment[offsetConstraint.variable].Uint64()), min, blockNumber)) {
+				return st.BlockContext{}, ErrUnsatisfiable
+			}
+			// else the existing value is already where we want it.
+		}
```
[the remark this answers is not in the document]
Comment (facuMH) [reply go/ct/gen/block_context.go:89]: I don't think so because `cur` could be a different number within the desired in/out of range condition. Maybe the comments should 
```
// If the variable is already bound, check if its value and the offset are both within the range of the block number            // or both out of the range. If neither is true, return an error.
```

File: go/ct/gen/block_context.go [hunk 9]
```diff
@@ -13,43 +13,153 @@
...
+	}
+
+	// generate block number
+	blockNumber, err = blockNumberRangeSolver.Generate(rnd)
 	if err != nil {
 		return st.BlockContext{}, err
 	}
+
+	variableRangeSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
+	baseSolver := NewRangeSolver(math.MinInt64, int64(blockNumber))
```
Comment (LuisPH3) [inline go/ct/gen/block_context.go:119]: Construction of this solver is trivial, I would construct a new one on each iteration.

File: go/ct/gen/block_context.go [hunk 10]
```diff
@@ -13,43 +13,153 @@
...
+	// for all non bound relevante variables, assign them a value based on the constraints.
+	for i, offsetConstraint := range b.variablesOffsetConstraints {
+		if _, isBound := assignment[offsetConstraint.lower.variable]; !isBound {
+			variableRangeSolver.Restore(baseSolver)
+			if difference := offsetConstraint.lower.offset - offsetConstraint.upper.offset; difference == 2 || difference == -2 {
+				// if the difference between the two offsets is 2, then we can only generate a fix value.
+				variableRangeSolver.AddEqualityConstraint(offsetConstraint.lower.offset - 1)
```
Comment (LuisPH3) [inline go/ct/gen/block_context.go:126]: lower -1 ? is this out of bounds? 
should it be lower +1 or upper -1 ?
I may be wrong.....

File: go/ct/gen/block_context.go [hunk 11]
```diff
@@ -13,43 +13,153 @@
...
+			} else if offsetConstraint.lower.offset > 1 && offsetConstraint.upper.offset < 256 {
+				// if lower offset is greater than 1 and upper is less than 256 we generate in range.
+				variableRangeSolver.AddLowerBoundary(offsetConstraint.upper.offset)
+				variableRangeSolver.AddUpperBoundary(offsetConstraint.lower.offset)
+			} else {
```
Comment (LuisPH3) [inline go/ct/gen/block_context.go:131]: There is something fishy in this code. The member `kind` of the `variableInequality` is never accessed.
I am not sure if this does the right thing for all scenarios, since the value is never used.  It is quite complicated as well. 

I presume this code could be simplified with the help of some member function in the `variableInequality`, and/or in the `constraintPair`, where get the bounded range.
Such a function would be much easier to unit test as well.  And  then the block in the loop turns into:
```
if _, isBound := assignment[offsetConstraint.lower.variable]; !isBound {
     newValue, err := offsetConstraint.Generate(rnd)
```

File: go/ct/gen/block_context.go [hunk 12]
```diff
@@ -13,43 +13,153 @@
...
+			}
+			newValue, err := variableRangeSolver.Generate(rnd)
+			if err != nil {
+				return st.BlockContext{}, err
+			}
+			assignment[offsetConstraint.lower.variable] = common.NewU256(uint64(int64(blockNumber) - newValue))
+		} else {
+			// if the variable is bound, then we need to check that it does not clash with any other constraint.
+			for _, previousConstraint := range b.variablesOffsetConstraints[:i] {
+				if previousConstraint.lower.variable == offsetConstraint.lower.variable {
```
Comment (LuisPH3) [inline go/ct/gen/block_context.go:151]: Same applies here, a member function in the `constraintPair` would make this test easier.

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
I saw something I did not complete understood:  Long story short, we have hard-coded some logic that should be computed based on the `kind` member of `variableInequality`:

- Your tests seem to be testing the relevant value generation
- Nevertheless, the `kind` member is never accessed, 

This makes me think that the arithmetic is hard-coded into the generator, assigning semantics to lower/upper members, this is not bad, but if the code is complete and working we could remove the `variableInequality` or use it instead:

I think the problem comes from the pair lower, and upper bounds, and assigning semantics to them. If instead you have an array of `inequalityType` and a boolean op, the list could define the range (or out of range)  we are looking for: 

- `[ (x > 0)  ,  (x  < 266)] +  AND=> in range`
- `[ (x < 1) , (x  > 265)]  + OR=> or range`

Anyway, I am pretty sure that @HerbertJordan has a better overview about how this system should be used.

File: go/ct/gen/block_context.go (hunk 10 above)
Comment (facuMH) [reply go/ct/gen/block_context.go:126]: short answer is `lower.offset` is not the same as "lower bound". But let's discuss this offline and make it clear. Maybe the sintax or naming should be changed.

File: go/ct/gen/block_context.go (hunk 11 above)
Comment (LuisPH3) [reply go/ct/gen/block_context.go:113]: As discussed, here we are dealing with the tradeoff between implementing a generator that can deal with inequalities vs the ad-hoc implementation for this routine. 
because of time constraints, we leave the ad-hoc implementation

File: go/ct/gen/block_context.go [hunk 13]
```diff
@@ -13,43 +13,85 @@
...
+		variablesOffsetConstraints: []variableOffset{}}
 }
 
-func (*BlockContextGenerator) Generate(rnd *rand.Rand, revision common.Revision) (st.BlockContext, error) {
+func isWithin256Range(offset int64, min, max uint64) bool {
+	val := max - uint64(offset)
+	return min <= val && val < max
+}
+
+func (b *BlockContextGenerator) Generate(assignment Assignment, rnd *rand.Rand, revision common.Revision) (st.BlockContext, error) {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply go/ct/gen/block_context.go:44]: after a discussion offline, we decided to first check for any bounded variable before the call to the block context generator, and if there are any, constraint the generation of the `BlockNumber` to that pre-bounded variable (as well as the current revision). And it only becomes an unsatisfiable condition if `a+256` is still less than the first block number of the current revision, or `a-1` is greater than the last block number of the current revision.

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/gen/block_context.go [hunk 14]
```diff
@@ -13,43 +13,164 @@
...
+			newValue, err := variableRangeSolver.Generate(rnd)
+			if err != nil {
+				return st.BlockContext{}, err
+			}
+			assignment[offsetConstraint.variable] = common.NewU256(uint64(int64(blockNumber) - newValue))
+		} else {
+			// if the variable is bound, then we need to check that it does not clash with any other constraint.
+			for _, previousConstraint := range b.variablesOffsetConstraints[:i] {
+				if previousConstraint.variable == offsetConstraint.variable {
+
```
Comment (LuisPH3) [inline go/ct/gen/block_context.go:146]: Can you explain to me why is it not enough with testing that the assigned value is in or out of the constraint range? 
as I understand, both B and the offset variable ($a) have been assigned at this point, and the only reason why it would not be satisfiable is if $a is not in the constraint, right?

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/gen/block_context.go (hunk 14 above)
Comment (facuMH) [reply go/ct/gen/block_context.go:146]: discusses offline. We agreed on a simpler way to check this.

**Review by HerbertJordan — COMMENTED** [review]
I have reviewed everything but the generator by now. Since it is already taking me a long time, I am submitting the current comments I have so fare. I still need to look into the generator.

**Comment by facuMH** [comment]
This PR got too big and mixed files and issues. It will be split into several smaller PRs, to facilitate review and understanding.
